File: Python3.x/202-Happy Number.py
# Runtime: 48 ms, faster than 38.12% of Python3 online submissions for Happy Number.
# Memory Usage: 13.4 MB, less than 5.04% of Python3 online submissions for Happy Number.
class Solution:
    def isHappy(self, n: int) -> bool:
        mem = set()
        while n != 1:
            n = sum([int(i) ** 2 for i in str(n)])
            if n in mem: return False
            else: mem.add(n)
        else: return True

File: Algorithms/C++/202-Happy_Number.cpp
// Runtime: 4 ms, faster than 66.56% of C++ online submissions for Happy Number.
// Memory Usage: 8.7 MB, less than 34.61% of C++ online submissions for Happy Number.
class Solution {
public:
    bool isHappy(int n) {
        set<int> mem;
        while (n != 1) {
            string s = to_string(n);
            int new_n = 0;
            for (char c : s)
                new_n += ((int)c - 48) * ((int)c - 48);
            if (mem.find(new_n) != mem.end())
                return false;
            else
                mem.insert(n);
            n = new_n;
        }
        return true;
    }
};
